accepted/engine-host-functions.md

```diff
@@ -0,0 +1,251 @@
+ 
+Given the nature of `Store` isolation in Wasmtime, this means host function definitions must be recreated for each and every 
+request handled by the service.
+ 
+If instead host functions could be defined for an `Engine` (in addition to `Store`), then the repeated effort to define the 
+host functions would be eliminated and the time it takes to create an instance to handle a request would therefore be 
+noticeably reduced.
+ 
+# Proposal sketch
+[proposal]: #proposal
+ 
+As `Func` is inherently tied to a `Store`, it cannot be used to represent a function associated with an `Engine`.
+ 
+This RFC proposes not having an analogous type to represent these functions; instead, users will define the function via a 
+method on `Engine`, but use `Store` (or `Linker`) to get the `Func` representing the function.
+ 
+## Overview example
+ 
+A simple example that demonstrates defining a host function in an `Engine`:
+ 
+```rust
+let engine = Engine::default();
+engine.wrap_function("", "hello", |caller: Caller<'_>, ptr: i32, len: i32| {
+    let mem = caller.get_export("memory").unwrap().into_memory().unwrap();
+    println!(
+        "Hello {}!",
+        std::str::from_utf8(unsafe{ &mem.data_unchecked()[ptr..ptr + len] }).unwrap()
+    );
+})?;
+ 
+let module = Module::new(
+    &engine,
+    r#"
+    (module
+        (import "" "hello" (func $hello (param i32 i32)))
+        (memory (export "memory") 1)
+        (func (export "run")
+            i32.const 0
+            i32.const 5
+            call $hello
+        )
+        (data (i32.const 0) "world")
+    )
+    "#,
+)?;
+ 
+// Note: given that `Module` is `Sync + Send`, this could be on a request handling thread
+// As nothing is defined locally in the `Linker`, the engine definitions will
+// be used for satisfying imports
+let store = Store::new(module.engine());
+let linker = Linker::new(&store);
+ 
+let instance = linker.instantiate(&module)?;
+let run = instance
+    .get_export("run")
+    .unwrap()
+    .into_func()
+    .unwrap()
+    .get0::<()>();
+ 
+run()?;
+```
+ 
+## Changes to `wasmtime::Engine`
+ 
+This RFC proposes that the following methods be added to `Engine`:
+ 
+```rust
+/// Defines a function for the `Engine` for the given callback.
+///
+/// Use `Store::get_engine_function` to get a `Func` representing the function.
+///
+/// Note that the implementation of `func` must adhere to the `ty`
+/// signature given, error or traps may occur if it does not respect the
+/// `ty` signature.
+///
+/// Additionally note that this is quite a dynamic function since signatures
+/// are not statically known. For performance reasons, it's recommended
+/// to use [`Engine::wrap_function`] if you can because with statically known
+/// signatures the engine can optimize the implementation much more.
+pub fn define_function(
+    &self,
+    module: &str,
+    name: &str,
+    ty: FuncType,
+    func: impl Fn(Caller<'_>, &[Val], &mut [Val]) -> Result<(), Trap> + 'static,
```

Comment:
Oh I also think that this is going to need `Send` and `Sync` bounds to keep `Engine` both `Send` and `Sync` (and that'd also highlight that as an engine-global thing the function may get invoked concurrently)

Reply:
Most definitely will. *shakes fist at Markdown not compiling his code*